src/natcap/invest/urban_nature_access.py

```diff
@@ -347,328 +504,647 @@ def execute(args):
+        greenspace_reclassification_task = graph.add_task(
+            _reclassify_greenspace_area,
+            kwargs={
+                'lulc_raster_path': file_registry['masked_lulc'],
+                'lulc_attribute_table': args['lulc_attribute_table'],
+                'target_raster_path': greenspace_pixels_path,
+            },
+            target_path_list=[greenspace_pixels_path],
+            task_name='Identify greenspace areas',
+            dependent_task_list=[lulc_mask_task]
+        )
 
-    # This is "SUP_DEMi" from the user's guide
-    greenspace_supply_demand_task = graph.add_task(
-        pygeoprocessing.raster_calculator,
-        kwargs={
-            'base_raster_path_band_const_list': [
-                (file_registry['greenspace_budget'], 1),
-                (file_registry['aligned_population'], 1)
-            ],
-            'local_op': _greenspace_supply_demand_op,
-            'target_raster_path': (
-                file_registry['greenspace_supply_demand_budget']),
-            'datatype_target': gdal.GDT_Float32,
-            'nodata_target': FLOAT32_NODATA
-        },
-        task_name='Calculate per-capita greenspace supply-demand',
-        target_path_list=[file_registry['greenspace_supply_demand_budget']],
-        dependent_task_list=[
-             per_capita_greenspace_budget_task,
-             population_alignment_task,
-        ])
+        decayed_population_in_group_paths = []
+        decayed_population_in_group_tasks = []
+        for pop_group in split_population_fields:
+            search_radius_m = search_radii_by_pop_group[pop_group]
+            decayed_population_in_group_path = os.path.join(
+                intermediate_dir,
+                f'decayed_population_in_{pop_group}{suffix}.tif')
+            decayed_population_in_group_paths.append(
+                decayed_population_in_group_path)
+            decayed_population_in_group_tasks.append(graph.add_task(
+                _convolve_and_set_lower_bound,
+                kwargs={
+                    'signal_path_band': (
+                        proportional_population_paths[pop_group], 1),
+                    'kernel_path_band': (
+                        kernel_paths[search_radius_m], 1),
+                    'target_path': decayed_population_in_group_path,
+                    'working_dir': intermediate_dir,
+                },
+                task_name=f'Convolve population - {search_radius_m}m',
+                target_path_list=[decayed_population_in_group_path],
+                dependent_task_list=[
+                    kernel_tasks[search_radius_m],
+                    proportional_population_tasks[pop_group]]
+            ))
+
+        sum_of_decayed_population_path = os.path.join(
+            intermediate_dir,
+            f'decayed_population_all_groups{suffix}.tif')
+        sum_of_decayed_population_task = graph.add_task(
+            ndr._sum_rasters,
```

Comment:
Interesting! I could see this getting refactored when some other proposals in the pipeline get implemented.

Reply:
For sure!  I betcha there are a couple functions like this that could be abstracted out.  Basic arithmetic across a raster stack in particular (e.g. adding up a stack of rasters) feels like something we may want to handle with a spatial case for memory efficiency.